### PR TITLE
DBZ-6845 Add configurable property for Cloud Spanner Emulator host.

### PR DESCRIPTION
**Pull request by ShuranZhang**

### CODE REVIEW

File: src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java [hunk 1]
```diff
@@ -176,6 +177,14 @@ public abstract class BaseSpannerConnectorConfig extends CommonConnectorConfig {
             .withImportance(Importance.LOW)
             .withDescription("Spanner host");
 
+    public static final Field SPANNER_EMULATOR_HOST = Field.create(GCP_SPANNER_EMULATOR_HOST_PROPERTY_NAME)
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java:180]: Could you please change the property name not to contain the emulator in it, ideally it should be `gcp.spanner.host` and the word emulator should be remove from property names too.
We have a prcedent for it in PubSub sink - https://debezium.io/documentation/reference/2.3/operations/debezium-server.html#address
And only documentation should highlight that this is meant for emulator.

**Review by jpechane — CHANGES_REQUESTED** [review]

File: src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java [hunk 2]
```diff
@@ -7,51 +7,99 @@
...
+                "gcp.spanner.instance.id", new ConfigValue("gcp.spanner.instance.id", "kafka-connector", new ArrayList<>(), new ArrayList<>()),
+                "gcp.spanner.database.id", new ConfigValue("gcp.spanner.database.id", "kafkaspan", new ArrayList<>(), new ArrayList<>()),
+                "gcp.spanner.emulator.host", new ConfigValue("gcp.spanner.emulator.host", "http://localhost:9010", new ArrayList<>(), new ArrayList<>()));
 
         ConfigurationValidator.ValidationContext validationContext = new ConfigurationValidator.ValidationContext(configuration, configValueMap);
 
         ConnectionValidator connectionValidator = spy(ConnectionValidator.withContext(validationContext));
-        // doNothing().when(connectionValidator).tryConnectByCredentialsJsonPath();
+        connectionValidator.validate();
+        Assertions.assertEquals(false, connectionValidator.isSuccess());
```
Comment (nancyxu123) [inline src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java:59]: Shouldn't this be true?

File: src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java [hunk 3]
```diff
@@ -7,51 +7,99 @@
...
+                "gcp.spanner.instance.id", new ConfigValue("gcp.spanner.instance.id", "kafka-connector", new ArrayList<>(), errMessages),
+                "gcp.spanner.database.id", new ConfigValue("gcp.spanner.database.id", "kafkaspan", new ArrayList<>(), new ArrayList<>()),
+                "gcp.spanner.credentials.path", new ConfigValue("gcp.spanner.credentials.path", "no_path", new ArrayList<>(), new ArrayList<>()),
+                "gcp.spanner.credentials.json", new ConfigValue("gcp.spanner.credentials.json", "{}", new ArrayList<>(), new ArrayList<>()));
+
+        ConfigurationValidator.ValidationContext validationContext = new ConfigurationValidator.ValidationContext(configuration, configValueMap);
+        ConnectionValidator connectionValidator = spy(ConnectionValidator.withContext(validationContext));
         connectionValidator.validate();
-        Assertions.assertEquals(isSuccess, connectionValidator.isSuccess());
+        Assertions.assertEquals(false, connectionValidator.isSuccess());
```
Comment (nancyxu123) [inline src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java:103]: Which property is missing here?

File: src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java (hunk 2 above)
Comment (ShuranZhang) [reply src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java]: Yes you are correct. I have updated the assertion here as well as the logic for credentials check. Now [`canValidate`](https://github.com/ShuranZhang/debezium-connector-spanner/blob/EmulatorTesting/src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java#L88) will only check for credentials if `isAgainstEmulator` is wrong.

File: src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java (hunk 3 above)
Comment (ShuranZhang) [reply src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java:102]: I have updated this test and it should be more clear now. As you can see the configuration only contain `gcp.spanner.database.id`, which misses other project/instance ids and the credentials.

File: src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java (hunk 1 above)
Comment (ShuranZhang) [reply src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java:180]: `gcp.spanner.host` and `gcp.spanner.emulator.host` are two different properties that need to be treated differently with `builder.setEmulatot(host)` and `builder.setEmulatorHost(emulatorHost)` when building a spanner connection(see this [file](https://github.com/ShuranZhang/debezium-connector-spanner/blob/EmulatorTesting/src/main/java/io/debezium/connector/spanner/db/DatabaseClientFactory.java#L61)). I'm not really sure what other word we can use to distinguish them if not by the configurable property name.
Or could you please suggest another word to use to distinguish between the two parameters other than "emulator" ? Thanks!
.

Comment (jpechane) [reply src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java:180]: @ShuranZhang ok, in that case could you use `createInternal` method that should make the config option `internal.gcp.spanner.emulator.host` and set it to invisible?

Comment (ShuranZhang) [reply src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java:180]: @jpechane Thanks for the suggestion! The point for making this Emulator host configurable is not for internal testing purpose. Running our connector against Emulator instance in test env is designed to be a customer facing new feature. Thus I don't feel it's very consistent to use a name starts with `internal.` and set it to invisible. In addition FYI,  we have a precedent for distinguishing between a regular Spanner host and the emulator host by using the word "emulator" for our apache beam connector [here](https://github.com/apache/beam/blob/4600517b4cd5100e9edd10b2ce9e224633fcb469/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerConfig.java#L60
). Let me know if you have other concerns, or you have other suggesting property name for `gcp.spanner.emulator.host`. Thanks!

**Review by jpechane — APPROVED** [review]
LGTM, @nancyxu123 feel free to merge when you are fine with the PR>

**Comment by nancyxu123** [comment]
> LGTM, @nancyxu123 feel free to merge when you are fine with the PR>

Thank you!

@ShuranZhang I was thinking we should probably hold off on submitting this until the ITs can pass?

**Review by nancyxu123 — APPROVED** [review]

**Comment by jpechane** [comment]
@ShuranZhang LGTM, thanks!
@nancyxu123 PLease feel free to merge the PR as soon as it is undrafted.

**Review by nancyxu123 — APPROVED** [review]

**Pull request merged**